wavefront/data_source_alert_test.go

```diff
@@ -21,6 +33,186 @@ func TestAccAlertIDRequired(t *testing.T) {
+	}
+
+	// Check if each flattened map matches the expected values
+	expected := []map[string]interface{}{
+		{
+			hostKey:   testHost1,
+			firingKey: testFiring1,
+		},
+		{
+			hostKey:   testHost2,
+			firingKey: testFiring0,
+		},
+	}
+
+	for i, expectedMap := range expected {
+		flattenedMap := flattened.([]map[string]interface{})[i]
+		for key, value := range expectedMap {
+			if flattenedMap[key] != value {
+				t.Errorf("Expected %s to be %v, but got %v", key, value, flattenedMap[key])
+			}
+		}
+	}
+}
+
+func TestSetAlertAttributes(t *testing.T) {
+	resource := dataSourceAlert()
+	var testAlertID = "test-id"
+	var testLinks = []string{testLink1, testLink2}
+	var testParameters = map[string]map[string]string{constantsKey: {testKey1: testVal1, testKey2: testVal2}}
+	var testAlertTriageDashboards = []wavefront.AlertTriageDashboard{
+		{
+			DashboardId: testDashboardID1,
+			Description: testDashboardDesc1,
+			Parameters:  testParameters,
+		},
+	}
+
+	// Create a sample schema.ReourceData
+	d := resource.TestResourceData()
+	alert := wavefront.Alert{
+		Name:                  testAlertName,
+		ID:                    &testAlertID,
+		RunbookLinks:          testLinks,
+		AlertTriageDashboards: testAlertTriageDashboards,
+	}
+
+	err := setAlertAttributes(d, alert)
+
+	// Check if there's no error
+	assert.NoError(t, err)
+
+	// Check if the fields in schema.ResourceData have been set correctly
+	assert.Equal(t, testAlertName, d.Get("name"))
+	assert.Equal(t, testAlertID, d.Get("id"))
+
+	// Check that the runbook_links field was set correctly
+	runbookLinksData := d.Get(runbookLinksKey).([]interface{})
+	assert.Equal(t, len(testLinks), len(runbookLinksData))
+	foundCount := 0
+	for _, linkData := range runbookLinksData {
+		for _, testLink := range testLinks {
+			if linkData == testLink {
+				foundCount++
+				break
+			}
+		}
+	}
+	assert.Equal(t, len(testLinks), foundCount)
+
+	// Check that the alert_triage_dashboards field was set correctly
+	alertTriageDashboardsData := d.Get(alertTriageDashboardsKey).([]interface{})
+	assert.Equal(t, len(testAlertTriageDashboards), len(alertTriageDashboardsData))
+	dashboardData := alertTriageDashboardsData[0].(map[string]interface{})
+	assert.Equal(t, testAlertTriageDashboards[0].DashboardId, dashboardData[dashboardIDKey].(string))
+	assert.Equal(t, testAlertTriageDashboards[0].Description, dashboardData[descriptionKey].(string))
+	parameterData := dashboardData[parametersKey].([]interface{})
+	assert.Equal(t, len(testParameters), len(parameterData))
+	constantsData := parameterData[0].(map[string]interface{})[constantsKey].(map[string]interface{})
+	assert.Equal(t, len(testParameters[constantsKey]), len(constantsData))
+	for key, value := range constantsData {
+		if key == testKey1 {
+			assert.Equal(t, testVal1, value)
+		}
+		if key == testKey2 {
+			assert.Equal(t, testVal2, value)
+		}
+	}
+}
+
```

Comment:
These are some meaty tests. I am definitely finding myself trying to reduce method complexity because I am imagining the really complex functions are going to have similarly massive unit tests. I think it's beyond the scope of your work and I am happy to see tests over not; but, it seems hard to call these "unit tests" due to their size.

I suppose it doesn't help that the data under test is a complex fairly complex structure.

Reply:
Yeah, I agree. A part of the problem is, as you say, that the data structures themselves are complex. But I will take another pass at this and see about reducing the complexity of these tests at least.